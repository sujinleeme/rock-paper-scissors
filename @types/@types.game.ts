export enum Choice {
  rock = 1,
  paper = 2,
  scissors = 3,
}

export enum Mode {
  humanVsComputer = 1,
  computers = 2,
}

export enum Winner {
  p1 = 'p1',
  p2 = 'p2',
  draw = 'draw',
}

export enum Record {
  win = 1,
  lose = 2,
  draw = 3,
}

export interface Player {
  name?: string
  color: string
  choices: Choice[]
  record: Winner[]
}

export interface FinalWinner {
  isDraw: boolean
  type: Winner
  name?: string
  color?: string
}

export type HashMap<T> = {
  [key in keyof T]: number
}
export interface GameState {
  isPlaying: boolean
  currentRound: number
  totalRounds: number
  isFinished: boolean
  possibleRounds: number[]
  mode?: Mode
  players: { p1: Player; p2: Player }
  winners: Winner[]
  score?: Map<Winner, number>
  winnerOfRound?: Winner
  finalWinner?: FinalWinner
}

export enum Size {
  'xs' = 0,
  'sm' = 1,
  'md' = 2,
}
