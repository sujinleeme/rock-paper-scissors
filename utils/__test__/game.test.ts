import { Choice } from '@types'

import { getRandomChoice, getWinner } from '../'

describe(`# getWinner`, () => {
  it.each`
    p1                 | p2
    ${Choice.rock}     | ${Choice.rock}
    ${Choice.scissors} | ${Choice.rock}
    ${Choice.rock}     | ${Choice.scissors}
  `(
    `returns winner when p1 choice is $p1 and p2 choice is $p2`,
    ({ p1, p2 }) => {
      expect(getWinner(p1, p2)).toMatchSnapshot()
    }
  )
})

describe(`# getRandomChoice`, () => {
  it('should return number betweeon 0 to 2', () => {
    const choice = getRandomChoice()
    const max = 2
    const isInRange = choice * (choice - max) <= 0
    expect(isInRange).toBe(true)
  })
})
