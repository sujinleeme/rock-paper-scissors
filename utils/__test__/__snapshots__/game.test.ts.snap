// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# getWinner returns winner when p1 choice is 0 and p2 choice is 0 1`] = `"draw"`;

exports[`# getWinner returns winner when p1 choice is 0 and p2 choice is 2 1`] = `"p1"`;

exports[`# getWinner returns winner when p1 choice is 2 and p2 choice is 0 1`] = `"p2"`;
